### PR TITLE
Fix #4102: Add Semaphore and use it in IOThrottler.

### LINKED ISSUE #4102: Implement java.util.concurrent.Semaphore and use it in IOThrottler
This is a follow-up to #4097 where we used an `AtomicInteger` as a poor-mans replacement for a `Semaphore` with `tryAcquire` (or maybe even `drainPermits`).

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Completing @yilinwei's work from #4308.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import org.junit.Assert._
+import org.junit.Test
+
+import org.scalajs.testsuite.utils.AssertThrows._
+
+class SemaphoreTest {
+
+  @Test def ctorUnfair(): Unit = {
+    val sem = new Semaphore(1)
+    assertFalse(sem.isFair)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala:27]: We usually use `()` when calling JDK methods, don't we?
```suggestion
    assertFalse(sem.isFair())
```
Same for `availablePermits()`

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+  @Test def ctorNegativePermits(): Unit = {
+    val sem = new Semaphore(-1)
+    assertEquals(-1, sem.availablePermits)
+    assertFalse(sem.tryAcquire())
+    sem.release()
+    assertEquals(0, sem.availablePermits)
+  }
+
+  @Test def drainNegative(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala:38]: We're missing a test of `drainPermits()` when the current available permits is positive, which is supposed to be the common case ;)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    assertEquals(3, sem.availablePermits)
+  }
+
+  @Test def releaseNegative(): Unit = {
+    val sem = new Semaphore(0)
+    assertThrows(classOf[IllegalArgumentException], sem.release(-1))
+  }
+
+  @Test def reducePermitsIntoNegative(): Unit = {
+    class ReducableSemaphore(permits: Int) extends Semaphore(permits) {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala:95]: Typo:
```suggestion
    class ReducibleSemaphore(permits: Int) extends Semaphore(permits) {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    assertFalse(sem.tryAcquire())
+    assertEquals(0, sem.availablePermits)
+
+    sem.reducePermits(2)
+    assertEquals(-2, sem.availablePermits)
+    assertFalse(sem.tryAcquire())
+
+    sem.release(3)
+    assertEquals(1, sem.availablePermits)
+    assertTrue(sem.tryAcquire())
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala:113]: This test is missing an `assertThrows` for calling `reducePermits` with a negative value.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    assertEquals(0, sem.availablePermits)
+  }
+
+  @Test def tryAcquire(): Unit = {
+    val sem = new Semaphore(1)
+    assertTrue(sem.tryAcquire())
+    assertEquals(0, sem.availablePermits)
+  }
+
+  @Test def tryAcquireNoneAvailable(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala:51]: Consider merging the tests that test the same method. For example, this one with `tryAcquire()` and `tryAcquireNegative()`.

File: javalib/src/main/scala/java/util/concurrent/Semaphore.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.util.concurrent
+
+import java.util.{Collection, Collections}
+
+class Semaphore(private[this] var permits: Int, fairness: Boolean) extends Serializable {
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/Semaphore.scala:17]: That's probably `scala.Serializable`. We need to explicitly import `java.io.Serializable`.

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by gzm0** [comment]
Updated. I think I have addressed all comments.

**Review by sjrd — APPROVED** [review]

**Pull request merged**